File: bad-codebase-python/app/models/customer.py
import sqlite3



class Customer:
    def __init__(self, row):
        self.id      = row['id']
        self.name    = row['name']
        self.email   = row.get('email')
        self.address = row.get('address')

    def get_display_name(self):
        # User.get_display_name() retorna 'Nome <email>'
        # Customer.get_display_name() retorna só o nome
        # código polimórfico quebra silenciosamente
        return self.name

    def is_high_value(self):
        # consulta o banco dentro do model — N+1 garantido
        conn = sqlite3.connect('database.sqlite')
        conn.row_factory = sqlite3.Row
        result = conn.execute(
            '''
            SELECT SUM(h.hours * b.hourly_rate) as total
            FROM billable_hours h
            JOIN billing_categories b ON h.category_id = b.id
            WHERE h.customer_id = ?
            ''',
            (self.id,)
        ).fetchone()
        conn.close()
        amount = result['total'] if result and result['total'] else 0
        return amount > 5000

    def get_month_total(self, month):
        conn = sqlite3.connect('database.sqlite')
        conn.row_factory = sqlite3.Row
        result = conn.execute(f"""
            SELECT SUM(h.hours * b.hourly_rate) as total
            FROM billable_hours h
            JOIN billing_categories b ON h.category_id = b.id
            WHERE h.customer_id = {self.id}
            AND strftime('%Y-%m', h.date_logged) = '{month}'
        """).fetchone()
        conn.close()
        return result['total'] if result and result['total'] else 0

File: bad-codebase/app/Models/Customer.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;

// Customer tem os mesmos campos que User mas os métodos se comportam diferente
// — um código que aceita User pode receber Customer e quebrar silenciosamente

// Um único Model gigante em vez de separar comportamentos em traits/interfaces

class Customer extends Model
{
    protected $table = 'customers';

    // $fillable incompleto — alguns campos salvos via DB::table()
    // não são protegidos aqui, criando inconsistência
    protected $fillable = ['name'];
    // email, address, description são salváveis via DB::table mas não via Eloquent
    // mass assignment protection quebrado pela metade

    // Sem relacionamentos definidos — joins são feitos com SQL cru no Controller
    // public function billableHours() { return $this->hasMany(BillableHour::class); }
    // (comentado porque "vai ser adicionado depois")

    // mas retorna formato diferente
    public function getDisplayName()
    {
        // User::getDisplayName() retorna "Nome <email>"
        // Customer::getDisplayName() retorna só o nome
        // código que chama getDisplayName() em qualquer "entidade" vai quebrar
        return $this->name;
    }

    public function isHighValue()
    {
        // consulta o banco dentro do model — N+1 garantido
        $total = \DB::select(
            'SELECT SUM(h.hours * b.hourly_rate) as total
             FROM billable_hours h
             JOIN billing_categories b ON h.category_id = b.id
             WHERE h.customer_id = ?',
            [$this->id]
        );
        $amount = $total[0]->total ?? 0;
        return $amount > 5000;
    }

    // Mesmo método que BillingService::calculateInvoice() — duplicação
    public function getMonthTotal($month)
    {
        $rows = \DB::select(
            "SELECT SUM(h.hours * b.hourly_rate) as total
             FROM billable_hours h
             JOIN billing_categories b ON h.category_id = b.id
             WHERE h.customer_id = {$this->id}
             AND DATE_FORMAT(h.date_logged, '%Y-%m') = '$month'"
        );
        return $rows[0]->total ?? 0;
    }
}
